### PR TITLE
Added PDF font switcher

### PR DESCRIPTION
**Pull request by HirokazuNishi**
Added PDF font switcher module for Japanese. It can generate  Invoice, Shipment, and Creditmemo with embedded Japanese font.

### CODE REVIEW

File: Kana/etc/fieldset.xml [hunk 1]
```diff
@@ -47,9 +47,11 @@
         <fieldset id="sales_convert_quote_address">
             <field name="firstnamekana">
                 <aspect name="to_order_address" />
+                <aspect name="to_customer_address" />
```
Comment (ishakhsuvarov) [inline Kana/etc/fieldset.xml]: Change set for this file seems to be out of scope for this Pull Request.

File: Pdf/Helper/Data.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * font file path
+     */
+    const FONT_DIR = 'lib/fonts/';
+
+    /**
+     * @var string
+     */
+    protected $_dir;
```
Comment (ishakhsuvarov) [inline Pdf/Helper/Data.php]: Current edition of [Magento Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html) suggests to use composition over inheritance, therefore it may be better to change modifier to `private`.

File: Pdf/Helper/Data.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+namespace MagentoJapan\Pdf\Helper;
+
+use Magento\Store\Model\ScopeInterface;
+use Magento\Framework\App\Helper\Context;
+
+/**
+ * Class Data
+ * @package MagentoJapan\Pdf\Helper
+ */
+class Data extends \Magento\Framework\App\Helper\AbstractHelper
```
Comment (ishakhsuvarov) [inline Pdf/Helper/Data.php]: Having `Helper` classes is not required and sometimes may even cause confusion for 3rd party developers to figure out the purpose of the class. I would suggest moving configuration reading logic to a separate model, with clear name reflecting it's purpose.

File: Pdf/Helper/Data.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+{
+
+    /**
+     * font file path
+     */
+    const FONT_DIR = 'lib/fonts/';
```
Comment (ishakhsuvarov) [inline Pdf/Helper/Data.php]: This constant might as well be made configurable via `di.xml` files.

File: Pdf/Helper/Data.php (hunk 2 above)
Comment (ishakhsuvarov) [inline Pdf/Helper/Data.php]: Adding underscore to private properties is more or less outdates convention. It's safe to use names without it.

File: Pdf/Model/Config/Source/JapaneseFont.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Class JapaneseFont
+ * @package MagentoJapan\Pdf\Model\Config\Source
+ */
+class JapaneseFont implements \Magento\Framework\Option\ArrayInterface
+{
+    /**
+     * font file path
+     */
+    const FONT_DIR = 'lib/fonts/';
```
Comment (ishakhsuvarov) [inline Pdf/Model/Config/Source/JapaneseFont.php]: There is already a similar constant used in the `Helper`. It would make sense to make it configurable from a single point.

File: Pdf/Model/Order.php [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    /**
+     * @param $object
+     * @param int $size
+     * @return mixed
+     */
+    protected function _setFontItalic($object, $size = 7)
+    {
+        if ($this->getHelper()->getJapaneseFontIsActive()) {
+            $fontpath = $this->getHelper()->getJapaneseFont();
```
Comment (ishakhsuvarov) [inline Pdf/Model/Order.php]: This part (following 3 lines) seems to be shared across multiple classes and methods. It makes sense to extract it.

File: Pdf/Test/Unit/Helper/DataTest.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?php
+namespace MagentoJapan\Pdf\Test\Unit\Helper;
+
+use Magento\Framework\TestFramework\Unit\Helper\ObjectManager;
+
+/**
+ * Class DataTest
+ * @package MagentoJapan\Pdf\Test\Unit\Helper
+ */
+class DataTest extends \PHPUnit_Framework_TestCase
```
Comment (ishakhsuvarov) [inline Pdf/Test/Unit/Helper/DataTest.php]: Unfortunately PHPUnit version used by Magento has been updated and this would not be compatible anymore with 2.2 and 2.3

**Review by ishakhsuvarov — COMMENTED** [review]
Some general comments:
* Which license is used by the included font files? Is it ok to distribute it?
* It's worth spending some additional and make all included code completely compatible with PHP 7.1+ standards.

**Comment by ishakhsuvarov** [comment]
Hi @HirokazuNishi 

Thank you for the PR. I've done a brief code review. Please feel free to take a look at the comments above. 

If you don't mind I would like to collaborate and push some of the fixes to your branch as well.

Thanks!

File: Pdf/Helper/Data.php (hunk 3 above)
Comment (vkublytskyi) [reply Pdf/Helper/Data.php]: Agree. @HirokazuNishi, to convert this class from helper to service class you may rename it to `MagentoJapan\Pdf\ModelConfig` and inject only scope config and dir reader as dependencies.

File: Pdf/Model/Config/Source/JapaneseFont.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * JapaneseFont constructor.
+     * @param \Magento\Framework\Module\Dir\Reader $dirReader
+     */
+    public function __construct(
+        \Magento\Framework\Module\Dir\Reader $dirReader
+    )
+    {
+        $this->_dir = $dirReader->getModuleDir('', 'MagentoJapan_Pdf') . '/' . self::FONT_DIR;
```
Comment (vkublytskyi) [inline Pdf/Model/Config/Source/JapaneseFont.php]: Inside constructor please only assign dependencies to private properties if something should be calculated it should be done when some method called

File: Pdf/Model/Creditmemo.php [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * @param $object
+     * @param int $size
+     * @return mixed
+     */
+    protected function _setFontRegular($object, $size = 7)
+    {
+        if ($this->getHelper()->getJapaneseFontIsActive()) {
```
Comment (vkublytskyi) [inline Pdf/Model/Creditmemo.php]: Such approach creates a lot of code duplication and if some 3rd party module has own pdf generated documents it will not be compatible.

We may try to create **after** plugin on `Magento\Framework\Filesystem\Directory\ReadInterface::getAbsolutePath` which will substitute result fort particular font paths. As we may limit check by `in_array` function call I believe there should not be dramatical performance degradation.

**Review by vkublytskyi — CHANGES_REQUESTED** [review]

**Comment by HirokazuNishi** [comment]
IPA font is distributed under [IPA font license](https://ipafont.ipa.go.jp/ipa_font_license_v1-html).

File: Pdf/Model/Creditmemo.php (hunk 9 above)
Comment (HirokazuNishi) [reply Pdf/Model/Creditmemo.php]: I tried to set plugin against Magento\Framework\Filesystem\Directory\ReadInterface, but Magento\Framework\Filesystem\Directory\ReadFactory has hardcoded class name for generating Read class instance. Seems it's impossible to set plugin against ReadInterface.

**Comment by ishakhsuvarov** [comment]
Static test failures regarding `Kana` module will be fixed in the scope of appropriate Pull Request
